aarch64: add necessary kvm ioctls

dianpopa opened the pull request:
`cargo test` is functional on aarch64 too.
Fixes https://github.com/firecracker-microvm/firecracker/issues/1040.

Signed-off-by: Diana Popa <dpopa@amazon.com>

andreeaflorescu commented:
Before merging this PR I would like to setup a CI for ARM as well to make sure that future PRs don't break the ARM build.

dianpopa commented:
> Before merging this PR I would like to setup a CI for ARM as well to make sure that future PRs don't break the ARM build.

Seems like a good idea. However, until then, any feedback is welcome.

andreeaflorescu commented on src/ioctls/device.rs:
```diff
@@ -0,0 +1,84 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
The current structure of kvm-ioctls is to have  3 sub-modules in the `ioctls` module that correspond to the KVM ioctls: system, vm and vcpu. When helper structures are required we can do one of two things:
- if they're used by more than one sub-module (system, vm or vcpu) the definition can lie in ioctls/mod.rs
- it they're used by only one sub-module than they should be added to the sub-module that uses them.

I believe `DeviceFd` is used only by `VmFd` and if this is not going to change in the future I say we place this code there rather than creating a new file so we can keep the initial structure.

dianpopa replied:
DeviceFd is not just a helper structure. KVM_CREATE_DEVICE (just like KVM_CREATE_VPCU and KVM_CREATE_VM) will give back an fd on which later you can call other kvm ioctls. Take a look at 
```
4.80 /KVM_GET_DEVICE_ATTR
Type: device ioctl, vm ioctl, vcpu ioctl
```
from the api.txt documentation. So `DeviceFd` corresponds to another type of ioctl, device ioctl (apart from vcpu and vm).
"I believe DeviceFd is used only by VmFd" -> DeviceFd will be used by regs and the global interrupt controller when all arm64 functionality will be merged.
I actually considered putting the code inside VMFd, however in VmFd is a lot of code already and it made it easier to add unit tests and add documentation putting it in a separate file.
If you still think it makes more sense to put it with VmFd, let me know.

andreeaflorescu replied:
Gotcha! Sounds good as it it. I checked the KVM API again, I missed the device ioctls entry before.

andreeaflorescu commented on src/ioctls/device.rs:
```diff
@@ -0,0 +1,84 @@
+}
+
+impl DeviceFd {
+    /// See the documentation for `KVM_SET_DEVICE_ATTR`.
```
Please add a short description of the ioctl: 
```
/// Sets a specified piece of device configuration and/or state.
///
/// See the documentation for `KVM_SET_DEVICE_ATTR`.
```

andreeaflorescu commented on src/ioctls/device.rs:
```diff
@@ -0,0 +1,84 @@
+        let kvm = Kvm::new().unwrap();
+        let vm = kvm.create_vm().unwrap();
+
+        let mut gic_device = kvm_bindings::kvm_create_device {
```
nit: can you import the structures and constants that you need from the kvm_bindings instead of using them with kvm_bindings::KVM_CREATE_DEVICE_TEST? The code is a bit hard to read.

andreeaflorescu commented on src/ioctls/device.rs:
```diff
@@ -0,0 +1,84 @@
+            fd: 0,
+            flags: kvm_bindings::KVM_CREATE_DEVICE_TEST,
+        };
+        // We expect this to fail. No VGIC_V3 on x86_64.
```
This is also expected to fail on ARM. Please update the comment.

andreeaflorescu commented on src/ioctls/vcpu.rs:
```diff
@@ -293,6 +300,37 @@ impl VcpuFd {
         Ok(())
     }
 
+    /// This initializes an ARM vCPU to the specified type with the specified features
```
Please format the doc comment with summary and description. In the description also add the KVM ioctl that the method is using (KVM_ARM_VCPU_INIT). Since this ioctl has a parameter also add the `# Arguments` section.

Same for `set_one_reg`.

andreeaflorescu commented on src/ioctls/vcpu.rs:
```diff
@@ -293,6 +300,37 @@ impl VcpuFd {
+    /// and resets the values of all of its registers to defaults.
+    #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+    pub fn vcpu_init(&self, kvi: &kvm_vcpu_init) -> Result<()> {
+        // safe becuase we allocated the struct and we know the kernel will read
```
s/becuase/because

andreeaflorescu commented on src/ioctls/vcpu.rs:
```diff
@@ -718,4 +756,38 @@ mod tests {
+        let mut reg_id: u64 = 0;
+
+        assert!(vcpu.set_one_reg(reg_id, data).is_err());
+        const PSTATE_REG_ID: u64 = 0x6030_0000_0010_0042;
```
Can you add a comment here regarding what are you trying to test and what is the `PSTATE_REG_ID`. This would be useful for people looking at the tests in the future.

andreeaflorescu commented on src/ioctls/vcpu.rs:
```diff
@@ -718,4 +756,38 @@ mod tests {
+
+        assert!(vcpu.set_one_reg(reg_id, data).is_err());
+        const PSTATE_REG_ID: u64 = 0x6030_0000_0010_0042;
+        vcpu.set_one_reg(PSTATE_REG_ID, data).expect("bla");
```
nit: We can change this expect into unwrap to get rid of the "bla".

andreeaflorescu commented on src/ioctls/vm.rs:
```diff
@@ -246,6 +255,28 @@ impl VmFd {
+
+    /// This queries the kernel for the preferred target CPU type.
+    #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+    pub fn get_preferred_target(&self, kvi: &mut kvm_vcpu_init) -> Result<()> {
```
We should try to name methods as close as possible to the KVM ioctls that they provide wrappers for. Usually the names of all methods strip the `KVM_` prefix and use lower case. For this one it would be arm_preferred_target.

dianpopa replied:
I find it ugly to put a cfg above this method that says it is arm specific and then name that method with an arm prefix. 
I think it is nice to keep that rule around but we can make exceptions when we feel there is a need to (see the register_irqfd function).

andreeaflorescu commented on src/ioctls/vm.rs:
```diff
@@ -358,11 +389,11 @@ mod tests {
+            assert!(vm_fd.register_irqfd(evtfd1, 4).is_ok());
+            assert!(vm_fd.register_irqfd(evtfd2, 8).is_ok());
+            assert!(vm_fd.register_irqfd(evtfd3, 4).is_ok());
+        }
```
With this change the following asserts fail for different reasons on arm and x86. I think it is worth mentioning why we expect the asserts to fail on each platform.

andreeaflorescu reviewed (commented):
General comments:
- Please update the coverage to 91.3. This is x86 specific unfortunately, but I added an issue to take care of it for ARM as well:  #9 .The coverage value is stored in tests/coverage.
- All doc comments (larger than one line) should be split in summary and description. Also, for methods that wrap KVM ioctls add a comment with the name of the ioctl that they provide wrappers for. If the methods have arguments, please add the `# Arguments` section to the comment.

andreeaflorescu commented:
@dianpopa I added a temporary fix for running cargo build with gnu/musl, and cargo test with gnu/musl on arm as well.

We still need to work on adding a coverage test for aarch64, but other than that we should be good to merge the PR once the tests pass.

The musl unit tests seem to fail for a reason that is unrelated to your PR. Do you happen to know why is there a linking problem? You can check the logs here: https://buildkite.com/rust-vmm/kvm-ioctls-ci/builds/108#55d3065d-1563-4a50-a0c9-2edd7c3f6afa

andreeaflorescu committed:
fixed link issues on aarch64 musl

The linker was unable to find __addtf3, __multf3 and __subtf3.

Added target-feature=+crt-static and link-arg=-lgcc as a temporary
workaround. This seems to be the accepted fix in the Rust community:
https://github.com/rust-lang-nursery/compiler-builtins/issues/201

A permanent fix is yet to be implemented in the Rust compiler.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu commented on src/ioctls/mod.rs:
```diff
@@ -1,12 +1,22 @@
-use kvm_bindings::kvm_run;
-#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
-use kvm_bindings::{kvm_cpuid2, kvm_cpuid_entry2};
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
nit: Adding the license is an unrelated change. Can you please add it in a new commit?

dianpopa replied:
Done, take a look.

andreeaflorescu commented on src/ioctls/device.rs:
```diff
@@ -0,0 +1,97 @@
+use sys_ioctl::ioctl_with_ref;
+
+/// Wrapper over the file descriptor obtained when creating an emulated device in the kernel.
+pub struct DeviceFd {
```
The structure need to be publicly exported in src/lib.rs to be accessible outside of the crate.

dianpopa replied:
Done, see lib.rs file.

dianpopa committed:
arch64: add necessary kvm ioctls

Tested that cargo test is functional on aarch64 too.

Signed-off-by: Diana Popa <dpopa@amazon.com>

dianpopa committed:
add license header for the "ioctls" module

Signed-off-by: Diana Popa <dpopa@amazon.com>
